# ===== Rules to count reads overlapping genes =================================


# Run featureCounts
rule featureCounts:
    input:
        expand(
            RESULTS + "/{sample}/{sample}_dedup.bam",
            sample = SAMS_UNIQ
        )
    output:
        RESULTS + "/" + PROJ + "_featureCounts.tsv"
    params:
        job_name = PROJ + "_featureCounts",
        args     = CMD_PARAMS["featureCounts"]
        gtf      = GTF
    resources:
        mem_gb = _get_mem()
    log:
        out = RESULTS + "/logs/" + PROJ + "_featureCounts.out",
        err = RESULTS + "/logs/" + PROJ + "_featureCounts.err"
    threads:
        16
    shell:
        """
        featureCounts \
            {params.args} \
            -F GTF \
            -a '{params.gtf}' \
            -o '{output}' \
            -T {threads} \
            {input}
        """


