# ===== Rules for intersecting reads with genomic regions ======================


# Intersect reads with metaplot regions
# these gene counts are used for the per-gene subsampling
# these are also used for generating plots without subsampling
use rule metaplot_beds as metaplot_beds_nosub with:
    input:
        RESULTS + "/{sample}/{sample}_shift.bed.gz"
    output:
        S    = meta_temp_fn(RESULTS + "/{sample}/metaplot_beds/{sample}_{region}_S.bed.gz"),
        AS   = meta_temp_fn(RESULTS + "/{sample}/metaplot_beds/{sample}_{region}_AS.bed.gz"),
        S_N  = meta_temp_fn(RESULTS + "/{sample}/metaplot_beds/{sample}_{region}_S_N.bed.gz"),
        AS_N = meta_temp_fn(RESULTS + "/{sample}/metaplot_beds/{sample}_{region}_AS_N.bed.gz")
    params:
        job_name = "{sample}_{region}_metaplot_beds",
        bed      = lambda wildcards: META_BEDS[wildcards.region]
    log:
        out = RESULTS + "/logs/{sample}_{region}_metaplot_beds.out",
        err = RESULTS + "/logs/{sample}_{region}_metaplot_beds.err"


