---
title: <p style="font-size:45px;">Spt5 mNET-seq</p>
author: "Ryan Sheridan"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      3
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    theme:          "cosmo"
    highlight:      "tango"

params:
  config: "PLOTS.yaml"
---

---

```{r "chunk opts"}

# Set default chunk options
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE,
  comment = "",
  echo    = FALSE,
  dpi     = 150,
  cache   = TRUE,
  cache.lazy = FALSE
)

```

```{r "plotting parameters"}

# Set plotting parameters
yml <- here::here(params$config)

plot_grps   <- yaml::read_yaml(yml)
plot_params <- plot_grps$params

plot_params$obj_dir <- file.path(res_dir, plot_params$obj_dir)
plot_params$tab_dir <- file.path(res_dir, plot_params$tab_dir)

knitr::knit(here::here(plot_params$template_dir, "setup.Rmd"), "")

```

```{r "load exons"}

ex_cols <- c(
  bed_cols,    "symbol",
  "gene_id",   "gene_name",
  "txn_id",    "ex_coords",
  "in_coords", "ex_len",
  "in_len",    "biotype",
  "seq"
)

exons <- read_tsv(
  str_c(list_dir, plot_params$exons),
  col_names = ex_cols
)

# Transcript info
# This must be passed separately since the exon bed file is generated from the
# gene pred file. There is no easy way to add the transcript_support_level to
# the gene pred file.
tx_cols <- c(
  bed_cols, "gene_id",
  "symbol", "biotype",
  "txn_id", "support"
)
  
tx_info <- here(list_dir, plot_params$trxn_info) %>%
  read_tsv(col_names = tx_cols) %>%
  select(txn_id, support)

```

```{r "1st exon length"}

# Identify first exon coordinates
# join with gene list to get gene coordinates
# only keep genes with 1st exon matching the gene TSS coordinate
exon_lens <- exons %>%
  left_join(genes_exons, by = c(gene_name = "name"), suffix = c("", ".gene")) %>%
  filter(
    !is.na(chrom.gene),
    score == 1,          # exon number, only keep 1st exon
    ((strand == "+" & start == start.gene) | (strand == "-" & end == end.gene))
  ) %>%
  
  # Add transcript support level
  left_join(tx_info, by = "txn_id") %>%
  
  select(
    chrom, start, end, score,
    strand, gene_name, txn_id, ex_coords,
    in_coords, ex_len, support
  ) %>%
  
  # Remove single exon genes
  # this is ~470 genes
  filter(!is.na(in_coords) & !is.na(support)) %>%
  
  # Use highest confidence transcript
  group_by(gene_name) %>%
  filter(support == min(support)) %>%
  ungroup()

# Transcripts to use for 5'ss metaplots
# some genes will have multiple transcripts
high_conf_txns <- unique(exon_lens$txn_id)

# Transcripts to use for quantifying exon/intron pausing
# exclude genes with multiple transcript IDs (~500)
ex_int_txns <- exon_lens %>%
  group_by(gene_name) %>%
  filter(n_distinct(txn_id) == 1) %>%
  pull(txn_id)

# Remove genes with multiple 1st exon coordinates
# this is a small number (~100)
exon_lens <- exon_lens %>%
  group_by(gene_name) %>%
  filter(n_distinct(ex_coords) == 1) %>%
  group_by(ex_len, .add = TRUE) %>%
  summarize(
    txn_id = first(txn_id),
    .groups = "drop"
  )

```

# Quality Control

`r knitr::knit_child(here(plot_params$template_dir, "qc-template.Rmd"))`

```{r "build chunks"}

# Create list of chunks
template <- here(plot_params$template_dir, "analysis-template.Rmd")

chunks <- names(plot_grps$samples) %>%
  map(~ knit_expand(template))

```

`r knit_child(text = chunks)`





```{r "ENSEMBL DATABASE", eval = FALSE}

# Gene coordinates
genes_all <- read_tsv(
  file.path(plot_params$ref_dir, plot_params$list_dir, plot_params$genes_all),
  col_names = bed_cols
)

# Create database from gtf
gtf <- "/home/rmsheridan/ref/annotation/GRCh38/gtf/Homo_sapiens.GRCh38.95.gtf"

ensdb <- here("ref/Homo_sapiens.GRCh38.95.sqlite")

if (!file.exists(ensdb)) {
  ensdb <- ensDbFromGtf(gtf = gtf, outfile = ensdb)
}

db <- EnsDb(ensdb)

```

```{r "FIG 1C", fig.width = 6, fig.height = 5, eval = FALSE}

# Load data for browser plots
bg_sams     <- c("+Spt5_r1", "+Spt5_r2")
bg_grps     <- c("Spt5-r1", "Spt5-r2")
bg_plt_grps <- c("Spt5", "Spt5")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)
pzones  <- .get_zones(bg_plt_grps, genes_all)

gene <- "BCAM"

create_browser(
  bgs        = bgs,
  gene       = gene,
  genes_df   = genes_all,
  genes_db   = db,
  pauses     = pauses,
  zones      = pzones,
  flank      = c(0.08, 0),
  track_clrs = bg_clrs,
  track_ttls = c("rep 1", "rep 2"),
  equal_axis = TRUE,
  pause_track_height = 0.1,
  gene_track_height  = 0.1,
  fontsize   = 18
)

```

