---
title: <p style="font-size:45px;">NET-seq analysis</p>
author: "Ryan Sheridan"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      3
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    theme:          "cosmo"
    highlight:      "tango"
params:
  samples: "~/Projects/find-pauses/SAMPLES.yaml"
  config:  "~/Projects/find-pauses/plots.yaml"
---

---

```{r "chunk opts"}

# Set default chunk options
knitr::opts_chunk$set(
  message    = FALSE, 
  warning    = FALSE,
  comment    = "",
  echo       = FALSE,
  dpi        = 150,
  cache      = FALSE,
  cache.lazy = FALSE
)

```

```{r "plotting parameters", include = FALSE}

# Sample groups
plot_grps <- yaml::read_yaml(here::here(params$samples))

# Plotting parameters
plot_params <- yaml::read_yaml(here::here(params$config))

plot_params$obj_dir <- file.path(res_dir, plot_params$obj_dir)
plot_params$tab_dir <- file.path(res_dir, plot_params$tab_dir)

knitr::knit(here::here(plot_params$template_dir, "setup.Rmd"), "")

```

```{r "load exons"}

ex_cols <- c(
  bed_cols,    "symbol",
  "gene_id",   "gene_name",
  "txn_id",    "ex_coords",
  "in_coords", "ex_len",
  "in_len",    "biotype",
  "seq"
)

exons <- read_tsv(
  str_c(list_dir, plot_params$exons),
  col_names = ex_cols
)

# Transcript info
# This must be passed separately since the exon bed file is generated from the
# gene pred file. There is no easy way to add the transcript_support_level to
# the gene pred file.
tx_cols <- c(
  bed_cols, "gene_id",
  "symbol", "biotype",
  "txn_id", "support"
)
  
tx_info <- here(list_dir, plot_params$trxn_info) %>%
  read_tsv(col_names = tx_cols) %>%
  dplyr::select(txn_id, support)

```

```{r "1st exon length"}

# Identify first exon coordinates
# join with gene list to get gene coordinates
# only keep genes with 1st exon matching the gene TSS coordinate
exon_lens <- exons %>%
  left_join(genes_exons, by = c(gene_name = "name"), suffix = c("", ".gene")) %>%
  dplyr::filter(
    !is.na(chrom.gene),
    score == 1,          # exon number, only keep 1st exon
    ((strand == "+" & start == start.gene) | (strand == "-" & end == end.gene))
  ) %>%
  
  # Add transcript support level
  left_join(tx_info, by = "txn_id") %>%
  
  dplyr::select(
    chrom, start, end, score,
    strand, gene_name, txn_id, ex_coords,
    in_coords, ex_len, support
  ) %>%
  
  # Remove single exon genes
  # this is ~470 genes
  dplyr::filter(!is.na(in_coords) & !is.na(support)) %>%
  
  # Use highest confidence transcript
  group_by(gene_name) %>%
  dplyr::filter(support == min(support)) %>%
  ungroup()

# Transcripts to use for 5'ss metaplots
# some genes will have multiple transcripts
high_conf_txns <- unique(exon_lens$txn_id)

# Transcripts to use for quantifying exon/intron pausing
# exclude genes with multiple transcript IDs (~500)
ex_int_txns <- exon_lens %>%
  group_by(gene_name) %>%
  dplyr::filter(n_distinct(txn_id) == 1) %>%
  pull(txn_id)

# Remove genes with multiple 1st exon coordinates
# this is a small number (~100)
exon_lens <- exon_lens %>%
  group_by(gene_name) %>%
  dplyr::filter(n_distinct(ex_coords) == 1) %>%
  group_by(ex_len, .add = TRUE) %>%
  summarize(
    txn_id = dplyr::first(txn_id),
    .groups = "drop"
  )

```

# Quality Control

`r knitr::knit_child(here(plot_params$template_dir, "qc-template.Rmd"))`

```{r "build chunks"}

# Create list of chunks
template <- here(plot_params$template_dir, "analysis-template.Rmd")

chunks <- names(plot_grps$SAMPLES) %>%
  map(~ knit_expand(template))

```

`r knitr::knit_child(text = chunks)`

# Session info

```{r, echo = FALSE}

sessionInfo()

```
